### PR TITLE
CLN: Share code between _BytesTarFile and _BytesZipFile

### LINKED ISSUE #47136: BUG: Pandas should not swallow exceptions when close()ing a write handle
### Pandas version checks

- [X] I have checked that this issue has not already been reported.

- [X] I have confirmed this bug exists on the [latest version](https://pandas.pydata.org/docs/whatsnew/index.html) of pandas.

- [X] I have confirmed this bug exists on the main branch of pandas.

### Reproducible Example

```python
pd.to_pickle(large_python_object, '/mnt/some_nfs_mount')
```

### Issue Description

`IOHandles.close` (https://github.com/pandas-dev/pandas/blob/v1.4.2/pandas/io/common.py#L119) throws away any exceptions raised by calling close() on the handle it is managing. When writing to certain file systems (in particular NFS mounts), write errors caused by e.g. runnning out of disk space may not be raised at the point you actually do a `write()` but rather get deferred to the `close()`. Therefore, by ignoring exceptions raised by `close()` it's possible for a Pandas IO utility function to silently write a truncated file.

### Expected Behavior

Exceptions raised by `close()` should be propagated to the user for appropriate handling. You may optionally want to continue to ignore exceptions raised by handles open in read mode.

### Installed Versions

<details>
INSTALLED VERSIONS
------------------
commit           : 66e3805b8cabe977f40c05259cc3fcf7ead5687d
python           : 3.8.12.final.0
python-bits      : 64
OS               : Linux
OS-release       : 4.18.0-305.19.1.el8_4.x86_64
Version          : #1 SMP Wed Sep 15 15:39:39 UTC 2021
machine          : x86_64
processor        : x86_64
byteorder        : little
LC_ALL           : None
LANG             : en_GB.UTF-8
LOCALE           : en_GB.UTF-8

pandas           : 1.3.5
numpy            : 1.20.3
pytz             : 2021.3
dateutil         : 2.8.2
pip              : 21.2.4
setuptools       : 58.0.4
Cython           : None
pytest           : 6.2.5
hypothesis       : None
sphinx           : None
blosc            : None
feather          : None
xlsxwriter       : None
lxml.etree       : 4.5.1
html5lib         : 1.1
pymysql          : None
psycopg2         : None
jinja2           : 3.0.2
IPython          : 7.31.1
pandas_datareader: None
bs4              : 4.10.0
bottleneck       : 1.3.2
fsspec           : 2022.01.0
fastparquet      : None
gcsfs            : None
matplotlib       : 3.5.0
numexpr          : 2.8.1
odfpy            : None
openpyxl         : 3.0.9
pandas_gbq       : None
pyarrow          : 4.0.1
pyxlsb           : None
s3fs             : 2022.01.0
scipy            : 1.7.3
sqlalchemy       : None
tables           : None
tabulate         : None
xarray           : 0.20.1
xlrd             : 2.0.1
xlwt             : None
numba            : 0.54.1
</details>

### PR DESCRIPTION
**Pull request by twoertwein**
- [x] xref #47136
- [x] [Tests added and passed]
- [x] All [code checks passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#pre-commit).
- [x] Added an entry in the latest `doc/source/whatsnew/vX.X.X.rst` file if fixing a bug or adding a new feature.

Needs some more work (test case + whatsnew) but want to first check whether the CI is happy. `_BytesTarFile` caused ValueErrors (due to inheriting from two IO classes?).

Would also like to further clean up `_BytesTarFile` and `_BytesZipFile` but that might be a separate PR.

### CODE REVIEW

File: doc/source/whatsnew/v1.5.0.rst [hunk 1]
```diff
@@ -814,6 +814,7 @@ I/O
 - Bug in :func:`read_excel` when reading a ``.ods`` file with newlines between xml elements (:issue:`45598`)
 - Bug in :func:`read_parquet` when ``engine="fastparquet"`` where the file was not closed on error (:issue:`46555`)
 - :meth:`to_html` now excludes the ``border`` attribute from ``<table>`` elements when ``border`` keyword is set to ``False``.
+- Most I/O methods do no longer suppress ``OSError`` and ``ValueError`` when closing file handles (:issue:`47136`)
```
Comment (simonjayhawkins) [inline doc/source/whatsnew/v1.5.0.rst:817]: see https://github.com/pandas-dev/pandas/issues/47136#issuecomment-1140222578

Comment (twoertwein) [reply doc/source/whatsnew/v1.5.0.rst:817]: If it should go for 1.4, I can make a PR just for 1.4. Literally just removing the try-catch (1.4 doesn't have the problematic tar support; and the cleanup doesn't need to be part of 1.4 either).

Comment (jreback) [reply doc/source/whatsnew/v1.5.0.rst:817]: prob worth a targeted patch for 1.4

Comment (twoertwein) [reply doc/source/whatsnew/v1.5.0.rst:817]: opened #47160. Will move the whatsnew to 1.4.3 in this PR to keep the PRs in sync.

**Review by simonjayhawkins — CHANGES_REQUESTED** [review]
#47160 should be merged first

**Comment by twoertwein** [comment]
Will submit a bug-fix after this PR is merged.

**Review by jreback — APPROVED** [review]
ok let's merge this, then do the appropriate backport (also prob need a whats new for 1.5 / different one for 1.4.3)

**Comment by jreback** [comment]
can you add some minimal testing for this & whatsnew note

**Comment by twoertwein** [comment]
> ok let's merge this, then do the appropriate backport (also prob need a whats new for 1.5 / different one for 1.4.3)

It fixed a bug in the not-yet-released tar feature. Does this need a whatsnew "even better tar support ;)"?

Will add a test.

**Comment by jreback** [comment]
right the note is about the propagating closing errors (don't need anything for changes on unrejeased)

**Comment by twoertwein** [comment]
> right the note is about the propagating closing errors (don't need anything for changes on unrejeased)

I changed this PR to still catch the error (if used within `get_handle`, the test calls the class directly). A follow-up PR (for main+1.4) will then remove the try-catch. See https://github.com/pandas-dev/pandas/pull/47160#issuecomment-1140451661

**Comment by simonjayhawkins** [comment]
> I changed this PR to still catch the error

should probably update title.

**Review by simonjayhawkins — APPROVED** [review]

**Comment by simonjayhawkins** [comment]
> > I changed this PR to still catch the error
> 
> should probably update title.

@twoertwein can you do this and then merge when ready.

**Pull request merged**